temporal-sdk/src/test/java/io/temporal/workflow/childWorkflowTests/ChildWorkflowContinueAsNewCancellationScopeTest.java

```diff
@@ -0,0 +1,164 @@
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.workflow.childWorkflowTests;
+
+import static org.junit.Assert.assertTrue;
+
+import io.temporal.api.common.v1.WorkflowExecution;
+import io.temporal.api.enums.v1.EventType;
+import io.temporal.client.WorkflowFailedException;
+import io.temporal.client.WorkflowOptions;
+import io.temporal.client.WorkflowStub;
+import io.temporal.failure.CanceledFailure;
+import io.temporal.workflow.Async;
+import io.temporal.workflow.CancellationScope;
+import io.temporal.workflow.ChildWorkflowCancellationType;
+import io.temporal.workflow.ChildWorkflowOptions;
+import io.temporal.workflow.Promise;
+import io.temporal.workflow.Workflow;
+import io.temporal.workflow.shared.SDKTestWorkflowRule;
+import io.temporal.workflow.shared.TestActivities.TestActivitiesImpl;
+import io.temporal.workflow.shared.TestWorkflows.NoArgsWorkflow;
+import io.temporal.workflow.shared.TestWorkflows.TestWorkflowCancellationType;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class ChildWorkflowContinueAsNewCancellationScopeTest {
+  private static AtomicInteger count;
+  private static CountDownLatch latch = new CountDownLatch(1);
+
+  @Rule
+  public SDKTestWorkflowRule testWorkflowRule =
+      SDKTestWorkflowRule.newBuilder()
+          .setWorkflowTypes(TestParentWorkflowImpl.class, TestChildWorkflowImpl.class)
+          .setActivityImplementations(new TestActivitiesImpl())
+          .build();
+
+  @Test
+  public void testChildWorkflowContinueAsNewCancellationTest() throws InterruptedException {
+    count = new AtomicInteger(2);
+    WorkflowStub parentWorkflow =
+        testWorkflowRule.newUntypedWorkflowStub("TestWorkflowCancellationType");
+    WorkflowExecution parentExecution =
+        parentWorkflow.start(ChildWorkflowCancellationType.WAIT_CANCELLATION_COMPLETED);
+    Thread.sleep(500);
+    parentWorkflow.cancel();
+    try {
+      parentWorkflow.getResult(String.class);
+      Assert.fail("unreachable");
+    } catch (WorkflowFailedException e) {
+      assertTrue(e.getCause() instanceof CanceledFailure);
+    }
+    testWorkflowRule.assertHistoryEvent(
+        parentExecution, EventType.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_CANCELED);
+  }
+
+  @Test
+  public void testImposterWorkflowContinueAsNewCancellationTest() throws InterruptedException {
```

Comment:
@mfateev @vitarb, could you please take a one last look at this test? I feel like I'm so close to reproducing Wenquan's [scenario](https://github.com/temporalio/sdk-java/issues/501#issuecomment-862620117), but getting a InternalWorkflowTaskException.